Assorted FreeBSD fixes.

arrowd opened the pull request.

adriaandegroot commented on CMakeLists.txt:
```diff
@@ -235,6 +235,8 @@ add_subdirectory(components)
 add_subdirectory(services)
 add_subdirectory(src)
+
+enable_testing()
```
Alternately, one might `include(CTest)` and rely on the *BUILD_TESTING* option that that introduces. But that's nit-picking: this is a good change because otherwise you have tests that cannot be run.

arrowd replied:
That's exactly why I added it.

adriaandegroot commented on data/scripts/Xsession:
```diff
@@ -52,6 +52,9 @@ case $SHELL in
 esac
 
 [ -f /etc/xprofile ] && . /etc/xprofile
+if [ `uname` = "FreeBSD" ]; then
```
Is this worth checking FreeBSD-ness for? The other lines are all "does this file exist? if yes, source it" which would make sense here as well; if you have an `/usr/local/etc/profile` and you're on Linux (or Solaris) then that's probably a profile to use there as well.

adriaandegroot commented on data/scripts/Xsession:
```diff
@@ -79,7 +79,7 @@ if [ -d "$xsessionddir" ]; then
         script="$xsessionddir/$i"
         echo "Loading X session script $script"
-        if [ -r "$script"  -a -f "$script" ] && expr "$i" : '^[[:alnum:]_-]\+$' > /dev/null; then
+        if [ -r "$script"  -a -f "$script" ] && expr "$i" : '^[[:alnum:]_-]\{1,\}$' > /dev/null; then
```
For reference:
```
[adridg@beastie ~/src/git/sddm]$ expr "abab" : '[ab]\{1,\}'
4
[adridg@beastie ~/src/git/sddm]$ expr "abab" : '[ab]\+'
0
adridg@tuxy:~$ expr "abab" : '[ab]\{1,\}'
4
adridg@tuxy:~$ expr "abab" : '[ab]\+'
4
```

adriaandegroot commented on src/helper/HelperApp.cpp:
```diff
@@ -306,13 +306,17 @@ namespace SDDM {
         if (!authSuccessful) {
 #if defined(Q_OS_LINUX)
             updwtmpx("/var/log/btmp", &entry);
+#elif defined(Q_OS_FREEBSD)
```
I don't think this is correct: as commented in a2b14425acc3b5659f44830c3ffd1d8d7c8f4c6a, `pututxline(3)` writes to a file depending on *ut_type* in the entry, and there's no "failued logins" user accounting database. And `pututxline()` has already been called in L301, so at this point it's just writing duplicates.

arrowd replied:
I simply pulled that patch from our port: https://svnweb.freebsd.org/ports/head/x11/sddm/files/patch-src_helper_HelperApp.cpp?view=markup

If this is wrong, should I drop it from the port too?

adriaandegroot replied:
For full context, the code looks like this (after patching in packaging, from L301):

```
        if (!pututxline (&entry))
            qWarning() << "Failed to write utmpx: " << strerror(errno);
        endutxent();
  
        // append to failed login database btmp
        if (!authSuccessful) {
#if defined(Q_OS_FREEBSD)
            pututxline(&entry);
#else
            updwtmpx("/var/log/btmp", &entry);
#endif
        }

        // append to wtmp
        else {
#if defined(Q_OS_FREEBSD)
            pututxline(&entry);
#else
            updwtmpx("/var/log/wtmp", &entry);
#endif
        }
```

On L301, a call to `pututxline(&entry)` .. puts a utx entry. In the subsequent `if()`, there are two more calls to `pututxline(&entry)` which **don't** change the entry .. this is just duplicating entries in the one-and-only utmp database that is used. It's just .. useless, and you could `#if !defined(Q_OS_FREEBSD) ` around the entire `if()` from `// append to failed` to the end of `// append to wtmp` because it's not going to do anything.

adriaandegroot commented on src/helper/Backend.cpp:
```diff
@@ -70,6 +75,26 @@ namespace SDDM {
+            setclassenvironment(lc, pw, 1);     /* path variables */
+            setclassenvironment(lc, pw, 0);     /* non-path variables */
+            login_close(lc);
+            if (lc = login_getuserclass(pw)) {
```
Is it **only** possible to have a valid userclass setting, when there is already (also) a pwclass setting? If not, I'd structure this as:
```
if (lc = login_getXXclass(pw)) { ... ; login_close(lc); }
if (lc = login_getYYclass(pw)) { ... ; login_close(lc); }
```
Although then you'll need some kind of flag to indicate that `env.insert()` should be called (call it only if something was actually set).

arrowd replied:
Re-routing original author's comment on this:

> This is modeled after [r357563 of FreeBSD/base](https://svnweb.freebsd.org/base/head/usr.bin/env/env.c?r1=357563&r2=357562&pathrev=357563) (line 173). I am not sure here, but it is probably the case that there is always a default pwclass configured so that would be found anyway. Also, I think it is doubtful to allow logging in only with user settings.

adriaandegroot replied:
In downstream r527832 of the patch in question this was changed; my original question regarded the construction

```
if ((lc = login_getpwclass(pw)) != NULL) {
    ...
    if ((lc = login_getuserclass(pw)) != NULL) {
        ...
    }
}
```

Note the nested `if()`s; I think it would be possible to have a valid userclass without a valid pwclass. The revised downstream patch follows the un-nested structure.

arrowd replied:
>  I think it would be possible to have a valid userclass without a valid pwclass.

But as already noted, `/usr/bin/env` stops further processing if `login_getpwclass` returns NULL. I think this is because `login_getpwclass` falls back to `default` class if the `pw` points to a nonexistent record, so NULL returned means error state. I believe that nested structure should be preserved.

adriaandegroot commented on src/helper/Backend.cpp:
```diff
@@ -70,6 +75,26 @@ namespace SDDM {
+            login_close(lc);
+            if (lc = login_getuserclass(pw)) {
+                setclassenvironment(lc, pw, 1);
+                setclassenvironment(lc, pw, 0);
```
To me the documentation suggests that `setusercontext()` is a lower-level function, and I was going to suggest `setusercontext(lc, pw, -1, LOGIN_SETPATH | LOGIN_SETENV)`. But looking in `login_class.c`, it turns out that `setusercontext()` splits out calls into `setclassenvironment()` and others; it's probably worth a one-line comment like
```
// setclassenvironment() is the implementation inside setusercontext()
```

adriaandegroot commented on src/helper/Backend.cpp:
```diff
@@ -70,6 +75,26 @@ namespace SDDM {
+            if (lc)
+                login_close(lc);
+            /* copy all environment variables that are now set */
+            env.insert(QProcessEnvironment::systemEnvironment());
```
Are you sure the environment is clean(-ish) before calling `setclassenvironment()` and then inserting the whole thing here? The implementation in `setclassenvironment()` just calls `setenv()` so you're potentially pulling in the whole environment of SDDM.

arrowd replied:
I have pulled another revision of this patch from https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=244341

arrowd commented:
Bump.

arrowd commented:
Bump.

adriaandegroot commented:
LGTM, with one small reservation around https://github.com/sddm/sddm/pull/1234/files#diff-d7898f8a43d0f15c2f009d1c0ac2efd089ee8b297f38e8f12c1d93a78bfa4b5eR305 .. here, the structure **was**  an if and the body of the if- and else-blocks had `#ifdef ` in them -- effectively, only Linux had a body there. Now, all not-FreeBSD has the full body, so this might now annoy NetBSD since they get the Linux-specific if- and else-blocks now. But they can complain about that separately :)

arrowd commented:
> LGTM, with one small reservation around https://github.com/sddm/sddm/pull/1234/files#diff-d7898f8a43d0f15c2f009d1c0ac2efd089ee8b297f38e8f12c1d93a78bfa4b5eR305 .. here, the structure **was** an if and the body of the if- and else-blocks had `#ifdef ` in them -- effectively, only Linux had a body there. Now, all not-FreeBSD has the full body, so this might now annoy NetBSD since they get the Linux-specific if- and else-blocks now. But they can complain about that separately :)

Thanks for your review. Addressed your comments.
